Core: Support serializable and snapshot isolation for ReplacePartitions

szehon-ho opened the pull request:
* From https://mail-archives.apache.org/mod_mbox/iceberg-dev/202107.mbox/%3CCAMwmD1_c4u%2BwDSOLtScCyid%2BfHicBnKdmyTuKqFTG-uaE9eq6A%40mail.gmail.com%3E and offline discussion with Ryan
* This change will make "insert overwrite" with mode=dynamic support serializable isolation (a ValidationException will be thrown if a conflicting change is made) in the case where it is not selecting from the same table.
* This will catch cases where two insert overwrite happen at the same time, both can potentially succeed and both datas land.
* "insert overwrite" remains snapshot isolation in the case where the select is from the same table. For that to support serializable, we need to plug the snapshotId from initial Spark compile, which could be looked at on spark-extension side.

szehon-ho commented:
@rdblue  if you could look when you have time, thanks

rdblue commented on core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:
```diff
@@ -53,6 +61,45 @@ public ReplacePartitions validateAppendOnly() {
+  public void validate(TableMetadata currentMetadata) {
+    if (validateNoConflictingAppends) {
+      Expression conflictDetectionFilter;
+      if (writeSpec().fields().size() <= 0) {
```
You should be able to call `isUnpartitioned()`

szehon-ho replied:
Done

szehon-ho commented:
Updated, and also turned off serialzable isolation by default (can turn it on as well)

rdblue commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -35,6 +35,7 @@
  * changes.
  */
 public interface ReplacePartitions extends SnapshotUpdate<ReplacePartitions> {
+
```
Nit: unnecessary whitespace change.

szehon-ho replied:
Done

rdblue commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +50,34 @@
+   * @return this for method chaining
+   */
+  ReplacePartitions validateFromSnapshot(long snapshotId);
+
```
Nit: extra newline.

szehon-ho replied:
Done

rdblue commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +50,34 @@
+   * <p>
+   * This method should be called when the table is queried to determine which files to overwrite.
+   * If a concurrent operation commits a new file after the data was read and that file might
+   * contain rows matching the specified conflict detection filter, the overwrite operation
```
This references "the specified conflict detection filter" but there isn't a filter. Can you update this to state that conflicts are detected by checking for any files added to the partitions that are replaced?

szehon-ho replied:
You are right, I rewrote the comment

rdblue commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +50,34 @@
+   * <p>
+   * Calling this method with a correct conflict detection filter is required to maintain
+   * serializable isolation for eager partition overwrite operations. Otherwise, the isolation level
+   * will be snapshot isolation.
```
I don't think that this applies anymore.

szehon-ho replied:
Thanks, removed

rdblue commented on core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:
```diff
@@ -53,6 +61,45 @@ public ReplacePartitions validateAppendOnly() {
+        // Unpartitioned table, check against all files
+        conflictDetectionFilter = Expressions.alwaysTrue();
+      } else {
+        conflictDetectionFilter = deletedPartitions.stream().map(p -> {
```
I don't think it's necessary to convert the partition set to a filter. You should be able to just omit the filter when configuring the `ManifestGroup` and apply it to each data file returned by the group.

szehon-ho replied:
Done

rdblue commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -247,11 +247,39 @@ private ManifestFile copyManifest(ManifestFile manifest) {
+   * @param conflictDetectionFilter a partition expression used to find new conflicting data files
+   * @param caseSensitive whether expression evaluation should be case sensitive
+   */
+  protected void validateAddedFilesWithPartFilter(TableMetadata base, Long startingSnapshotId,
```
Rather than splitting the expression, I think you can write an alternative method that checks a `PartitionSet`.

szehon-ho replied:
Done

rdblue commented on core/src/test/java/org/apache/iceberg/TestReplacePartitions.java:
```diff
@@ -253,6 +253,189 @@ public void testValidationSuccess() {
+
+    AssertHelpers.assertThrows("Should reject commit with file matching partitions replaced",
+        ValidationException.class,
+        "Found conflicting files that can contain records matching true: [/path/to/data-b.parquet]",
```
"records matching true" is not the correct error message. Creating a new validation that uses a partition set would allow you to fix the error message easily.

szehon-ho replied:
Done

rdblue commented on core/src/test/java/org/apache/iceberg/TestReplacePartitions.java:
```diff
@@ -253,6 +253,189 @@ public void testValidationSuccess() {
+    validateManifestEntries(table.currentSnapshot().allManifests().get(1),
+        ids(replaceId),
+        files(FILE_E),
+        statuses(Status.DELETED));
```
It isn't clear to me why FILE_E was deleted. Is that in the same partition as FILE_A or FILE_B? If so, then we need to do something to make the test more readable so it is clear why this is correct.

szehon-ho replied:
Yea FILE_E is same partition as FILE_A.  I added a comment a few lines earlier explaining, will it suffice?

rdblue commented on core/src/test/java/org/apache/iceberg/TestReplacePartitions.java:
```diff
@@ -253,6 +253,189 @@ public void testValidationSuccess() {
+    validateManifestEntries(table.currentSnapshot().allManifests().get(2),
+        ids(replaceId),
+        files(FILE_A),
+        statuses(Status.DELETED));
```
Same with this test. It isn't clear why these deletes happened.

rdblue commented on core/src/test/java/org/apache/iceberg/TestReplacePartitions.java:
```diff
@@ -253,6 +253,189 @@ public void testValidationSuccess() {
+    AssertHelpers.assertThrows("Should reject commit with file matching partitions replaced",
+        ValidationException.class,
+        "Found conflicting files that can contain records matching true: [/path/to/data-a.parquet]",
+        replace::commit);
```
I think it is usually easier to read tests if you put the whole `replace` operation here, like this:

```java
        () ->
            unpartitioned.newReplacePartitions()
                ...
                .commit());
```

szehon-ho replied:
Good idea, done

szehon-ho reviewed (commented):
@rdblue thanks a lot for taking a look.  I rewrote it based on your suggestion (from using Expression to using a filter of partitionSet::contains).  Also added a toString method to PartitionSet to make the user-facing message.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:
```diff
@@ -20,11 +20,18 @@
 public class BaseReplacePartitions
     extends MergingSnapshotProducer<ReplacePartitions> implements ReplacePartitions {
+
+  private final PartitionSet deletedPartitions = PartitionSet.create(super.getSpecsById());
```
If there is a reference to other fields or methods, please do the initialization in the constructor.

szehon-ho replied:
Sure, done.

rdblue commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +49,28 @@
+   *
+   * @return this for method chaining
+   */
+  ReplacePartitions validateNoConflictingAppends();
```
After #3069 is in, I think this should also support `validateNoConflictingDeletes()` using the new validation in that PR.

rdblue commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -242,6 +245,53 @@ private ManifestFile copyManifest(ManifestFile manifest) {
+   * @param startingSnapshotId id of the snapshot current at the start of the operation
+   * @param partitionSet a set of partitions to check against, or none if check is to be against all files
+   */
+  protected void validateAddedDataFiles(TableMetadata base, Long startingSnapshotId,
```
Rather than using an `Optional<PartitionSet>` couldn't this use `validateAddedDataFiles` with filter `Expressions.alwaysTrue()`? That would make this a bit more straightforward. and avoid the `isPresent` check inside a loop (which probably isn't a good idea anyway).

szehon-ho replied:
Done.  only thing is it makes it back to a bad error message "Found conflicting files that can contain records matching true", if that's ok

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java:
```diff
@@ -138,7 +140,10 @@ BatchWrite asBatchAppend() {
+    String isolationLevelAsString = table.properties().getOrDefault(REPLACE_PARTITION_LEVEL,
+        REPLACE_PARTITION_LEVEL_DEFAULT);
+    IsolationLevel level = IsolationLevel.valueOf(isolationLevelAsString.toUpperCase(Locale.ROOT));
+    return new DynamicOverwrite(level, table.currentSnapshot().snapshotId());
```
You might consider doing this differently. Maybe accept a write option that sets the starting snapshot ID for now? Defaulting this to the current snapshot as of the time that the dynamic overwrite is created seems unreliable to me.

szehon-ho replied:
Got it, I implemented the suggestion (taking your example from https://github.com/apache/iceberg/pull/3069#discussion_r711803863), and added a Spark level unit test

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java:
```diff
@@ -260,16 +265,29 @@ public void commit(WriterCommitMessage[] messages) {
   }
 
   private class DynamicOverwrite extends BaseBatchWrite {
+
```
Nit: no need for a newline here.

szehon-ho replied:
Done

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java:
```diff
@@ -260,16 +265,29 @@ public void commit(WriterCommitMessage[] messages) {
 
+      if (isolationLevel == SERIALIZABLE) {
+        dynamicOverwrite.validateNoConflictingAppends();
+      }
```
Nit: please add a newline between control flow blocks and the next statement.

szehon-ho replied:
Sure, done.

szehon-ho commented:
@rdblue  thanks for the suggestions.  I put validate-from-snapshot as a SparkWriteOption, and added a Spark level test for it based on it, if you want to take another look.

szehon-ho commented:
This change now adds validation for conflicting delta files.  

Implementation note:  One option was to convert Expression => PartitionSet to re-use existing validation methods.  But as per earlier comment to keep PartitionSet, we go with new version of validateAddedFiles that takes in PartitionSet, we also now add new version of validateNoNewDeletesForDataFiles that takes PartitionSet.

On UX side, we have this enabled by user by setting following Spark write conf: validate-from-snapshot, as I understood from comment:  https://github.com/apache/iceberg/pull/3069#discussion_r711803863. This then triggers validation if dynamic.overwrite.isolation.level is serializable (the new default), else there is no validation.

szehon-ho commented:
@rdblue @aokolnychyi  this is ready for another round, when you guys have time, thanks!

aokolnychyi commented:
Let me take a look.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/DeleteFileIndex.java:
```diff
@@ -467,13 +474,25 @@ DeleteFileIndex build() {
-                  .caseSensitive(caseSensitive)
-                  .liveEntries()
-      );
+          manifest -> deleteManifestEntries(
```
This part becomes quite complicated. Would it make sense to add a method to `ManifestReader` directly? We have one that accepts an expression. We could have one more that accepts `PartitionSet`.

szehon-ho replied:
Yes refactored and cleaned it up

aokolnychyi commented on core/src/main/java/org/apache/iceberg/DeleteFileIndex.java:
```diff
@@ -467,13 +474,25 @@ DeleteFileIndex build() {
+          f -> partitionSet.contains(f.file().specId(), f.file().partition()));
     }
   }
+
```
nit: unnecessary change?

szehon-ho replied:
Reverted

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -263,6 +265,44 @@ private ManifestFile copyManifest(ManifestFile manifest) {
+
+    try (CloseableIterator<ManifestEntry<DataFile>> conflicts = conflictGroup.entries().iterator()) {
+      CloseableIterable<ManifestEntry<DataFile>> filtered = CloseableIterable.filter(conflictGroup.entries(),
+          f -> partitionSet.contains(f.file().specId(), f.file().partition()));
```
Can this be part of the predicate closure we pass to `filterManifestEntries`?

szehon-ho replied:
Yes, done

aokolnychyi commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -225,6 +225,9 @@ private TableProperties() {
   public static final String MIN_SNAPSHOTS_TO_KEEP = "history.expire.min-snapshots-to-keep";
   public static final int MIN_SNAPSHOTS_TO_KEEP_DEFAULT = 1;
 
+  public static final String DYNAMIC_OVERWRITE_ISOLATION_LEVEL = "write.dynamic.overwrite.isolation-level";
```
Are we sure we want to have a table property for this? Also, does this mean it will be a breaking change in the behavior given the currently proposed default?

I am not sure what's best yet but we can express the same things via data source options.

szehon-ho replied:
Yea, Flink does use this API.

One option is having sparkWriteConf, something like dynamic.overwrite.{table}.isolation.level?  Which then activate if paired with "validate-from-snapshot".  Not sure if that's the best way.  Didn't change yet so we can discuss

aokolnychyi replied:
What I mean is that we can use Dataset options for this. We anyway need to pass the snapshot ID that was scanned to make it work and the only way to do that is through options. Even if we have a property, we still need to set the option. I'd probably use options to begin with. I can be convinced otherwise, though.

aokolnychyi commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java:
```diff
@@ -270,6 +275,13 @@ public void commit(WriterCommitMessage[] messages) {
 
       ReplacePartitions dynamicOverwrite = table.newReplacePartitions();
 
+      String isolationLevelAsString = table.properties().getOrDefault(DYNAMIC_OVERWRITE_ISOLATION_LEVEL,
```
This can be part of `SparkWriteConf`.

szehon-ho commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -20,19 +20,23 @@
+ * correct and should be committed out regardless of other concurrent changes to the table.
+ * Alternatively, this API can be configured to validate that no new data or deletes
+ * have been applied since a snapshot id associated when this operation began.
+ * </p>
```
Removed, thanks

szehon-ho commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +53,19 @@
+   * Set the snapshot ID used in validations for this operation.
+   * <p>
+   * All validations will check changes after this snapshot ID. If this is not called or snapshot is set to an
+   * invalid snapshot id, validation will be skipped.
```
Done

szehon-ho commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +53,19 @@
+   * invalid snapshot id, validation will be skipped.
+   *
+   * This method should be called before this operation is committed.
+   * If a concurrent operation committed a data file or row delta after the given snapshot id
```
Done

szehon-ho commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +53,19 @@
+   * @param snapshotId a snapshot ID, it should be set to when this operation started to read the table.
+   * @return this for method chaining
+   */
+  ReplacePartitions validateFromSnapshot(long snapshotId);
```
Sure, rewrote to match

aokolnychyi replied:
Looks good now.

szehon-ho commented on core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:
```diff
@@ -22,12 +22,18 @@
 public class BaseReplacePartitions
     extends MergingSnapshotProducer<ReplacePartitions> implements ReplacePartitions {
+
+  private final PartitionSet deletedPartitions;
```
Good point, changed

szehon-ho commented on core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:
```diff
@@ -53,6 +60,24 @@ public ReplacePartitions validateAppendOnly() {
+  @Override
+  public void validate(TableMetadata currentMetadata) {
+    if (startingSnapshotId != null && startingSnapshotId != 0) {
+      if (dataSpec().isUnpartitioned()) {
```
This is tricky, but I am thinking we should validate everything, consistent with the current behavior as you mention in next comment (if unpartition, replace everything).  From the spark side it makes sense to me that a user running "insert overwrite" on a currently-unpartitioned table will expect to replace all the data whether they are in now-useless old partitioned folders or the new one, so the validation check should be the whole table.  (Unless we want to change base behavior)

szehon-ho commented:
Note: failures look unrelated (look like random hashing makes the file count different than expected in both cases)

szehon-ho commented:
FYI, filed https://github.com/apache/iceberg/pull/4033 and https://github.com/apache/iceberg/issues/4031 about the random failure of TestCopyOnWriteUpdate::testUpdateWithoutCondition

szehon-ho commented on core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:
```diff
@@ -22,12 +22,22 @@
     super(tableName, ops);
     set(SnapshotSummary.REPLACE_PARTITIONS_PROP, "true");
+    replacedPartitions = PartitionSet.create(ops.current().specsById());
+    this.startingSnapshotId = ops.current().currentSnapshot() == null ? 0 :
```
Makes sense, got rid of this default

szehon-ho commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +55,42 @@
+   */
+  ReplacePartitions validateFromSnapshot(long snapshotId);
+
+
```
Done

szehon-ho commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -20,19 +20,25 @@
+ * Alternatively, this API can be configured to validate that no new data or deletes
+ * have been applied since a snapshot ID associated when this operation began.
+ * This can be done by calling {@link #validateNoConflictingDeletes()}
+ * and {@link #validateNoConflictingData()}, and will ensure that no conflicting delta files or data files
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/DeleteFileIndex.java:
```diff
@@ -333,6 +334,7 @@ static Builder builderFor(FileIO io, Iterable<ManifestFile> deleteManifests) {
     private Expression partitionFilter = Expressions.alwaysTrue();
     private boolean caseSensitive = true;
     private ExecutorService executorService = null;
+    private PartitionSet partitionSet = null;
```
Yep, I usually do this but somehow did not here, done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -263,6 +264,44 @@ private ManifestFile copyManifest(ManifestFile manifest) {
+   * @param startingSnapshotId id of the snapshot current at the start of the operation
+   * @param partitionSet a set of partitions to check against
+   */
+  protected void validateAddedDataFiles(TableMetadata base, Long startingSnapshotId, PartitionSet partitionSet) {
```
Slightly bigger change, done, please help double-check

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -399,6 +438,35 @@ protected void validateNoNewDeleteFiles(TableMetadata base, Long startingSnapsho
+   * @param startingSnapshotId id of the snapshot current at the start of the operation
+   * @param partitionSet partition set to match against delete files
+   */
+  protected void validateNoNewDeleteFiles(TableMetadata base, Long startingSnapshotId, PartitionSet partitionSet) {
```
Done

szehon-ho commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestConflictValidation.java:
```diff
@@ -0,0 +1,121 @@
+    Assert.assertNotNull("Expected validation exception but none thrown", thrown);
+    Assert.assertTrue("Nested validation exception", thrown.getCause() instanceof ValidationException);
+    Assert.assertTrue(thrown.getCause().getMessage().contains(
+        "Cannot determine history between starting snapshot"));
```
@aokolnychyi after changing this path to use the common method MergingSnapshotProducer::validateAddedData/DeleteFiles, it will throw this kind of exception if user sets serializable-isolation but not the validate-from-snapshot (common method tries to get the history and it fails as there is no startingSnapshotId). Is this a good expected behavior?

aokolnychyi replied:
I think it is related to 0 snapshot ID used as default for `validate-from-snapshot-id`. I hope if we fix the default value, it will fix the error message.

szehon-ho replied:
You are right, this worked.

szehon-ho commented:
@aokolnychyi done if you want to take a look at another round, thanks!

aokolnychyi commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteOptions.java:
```diff
@@ -68,4 +68,10 @@ private SparkWriteOptions() {
+  // Identifies snapshot from which to start validating conflicting changes
+  public static final String VALIDATE_FROM_SNAPSHOT_ID = "validate-from-snapshot";
+
+  public static final String DYNAMIC_OVERWRITE_ISOLATION_LEVEL = "write.dynamic.overwrite.isolation-level";
```
We try to use short names for read and write options without dots to indicate namespace as opposed to table props. This option follows the table property convention.

What about making this property generic? We could call it `isolation-level` and then respect in DELETE, UPDATE, MERGE operations too (in a separate PR).

szehon-ho replied:
Good point, updated

aokolnychyi commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -236,4 +237,19 @@ public boolean useTableDistributionAndOrdering() {
+  public long validateFromSnapshotId() {
+    return confParser.longConf()
+        .option(SparkWriteOptions.VALIDATE_FROM_SNAPSHOT_ID)
+        .defaultValue(0)
```
I don't think 0 is a correct default. If we set it to 0, the validation will throw an exception that it cannot find a snapshot with ID = 0. Should this use `Long` and `parseOptional`? The place that calls it will have to be adapted too.

szehon-ho commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +55,41 @@
+  /**
+   * Set the snapshot ID used in validations for this operation.
+   *
+   * All validations will check changes after this snapshot ID. If this is not called, validation will occur
```
Changed

szehon-ho commented on core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:
```diff
@@ -53,6 +62,38 @@ public ReplacePartitions validateAppendOnly() {
+        validateAddedDataFiles(currentMetadata, startingSnapshotId, replacedPartitions);
+      }
+    }
+    if (validateNewDeleteFiles) {
```
Good catch, added the check and test of concurrently deleted data files as discussed.

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -263,6 +265,28 @@ private ManifestFile copyManifest(ManifestFile manifest) {
+            partitionSet,
+            Iterators.toString(Iterators.transform(conflicts, entry -> entry.file().path().toString())));
+      }
+    } catch (IOException e) {
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -285,22 +337,18 @@ protected void validateAddedDataFiles(TableMetadata base, Long startingSnapshotI
+    if (conflictDetectionFilter != null) {
+      conflictGroup = conflictGroup.filterData(conflictDetectionFilter);
     }
+    if (partitionSet != null) {
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -272,9 +296,37 @@ private ManifestFile copyManifest(ManifestFile manifest) {
+  }
+
+  /**
+   * Returns an iterable of files matching a filter have been added to the table since a starting snapshot.
```
Yea, changed the argument name

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -379,24 +423,63 @@ private void validateNoNewDeletesForDataFiles(TableMetadata base, Long startingS
+    // if there is no current table state, return empty delete file index
     if (base.currentSnapshot() == null || base.formatVersion() < 2) {
-      return;
+      return DeleteFileIndex.builderFor(ops.io(), Lists.newArrayList()).specsById(base.specsById()).build();
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -379,24 +423,63 @@ private void validateNoNewDeletesForDataFiles(TableMetadata base, Long startingS
+    if (partitionSet != null) {
+      deleteIndexBuilder = deleteIndexBuilder.filterPartitions(partitionSet);
+    }
+    if (dataFilter != null) {
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -379,24 +423,63 @@ private void validateNoNewDeletesForDataFiles(TableMetadata base, Long startingS
    * @param base table metadata to validate
    * @param startingSnapshotId id of the snapshot current at the start of the operation
-   * @param dataFilter an expression used to find new conflicting delete files
+   * @param dataFilter an expression used to filter new conflicting delete files
```
Basically i wrote 'filter' in all the new methods then came back and saw this existing method said 'find' and thought to change that.   Changed all of the new javadocs to 'find' instead.

szehon-ho commented on core/src/main/java/org/apache/iceberg/util/PartitionSet.java:
```diff
@@ -183,6 +184,26 @@ public boolean removeAll(Collection<?> objects) {
   }
 
+  @Override
+  public String toString() {
```
Should be the expected error messages in the new TestReplacePartition tests

szehon-ho commented:
@aokolnychyi thanks for the review, addressed the comments, when you another chance

szehon-ho commented:
@aokolnychyi about the open question of what to do for 'snapshot isolation', I gave it some thought:

From: https://www.postgresql.org/docs/13/transaction-iso.html#XACT-REPEATABLE-READ. (which according to it is almost like "Snapshot Isolation"):

> UPDATE, DELETE, SELECT FOR UPDATE, and SELECT FOR SHARE commands behave the same as SELECT in terms of searching for target rows: they will only find target rows that were committed as of the transaction start time. However, such a target row might have already been updated (or deleted or locked) by another concurrent transaction by the time it is found. In this case, the repeatable read transaction will wait for the first updating transaction to commit or roll back (if it is still in progress). If the first updater rolls back, then its effects are negated and the repeatable read transaction can proceed with updating the originally found row. But if the first updater commits (and actually updated or deleted the row, not just locked it) then the repeatable read transaction will be rolled back with the message

> ERROR:  could not serialize access due to concurrent update

> because a repeatable read transaction cannot modify or lock rows changed by other transactions after the repeatable read transaction began.

I think "insert overwrite" / ReplacePartitions is more like "Delete *" followed by "Insert * ".  So I think we may follow the steps here they describe here for DELETE and throw exception if any row we are to delete has been modified.  So here, if we see either of 1. delete file or 2. deleted data files, we throw the exception.

It also makes sense to me from the definition of "Snapshot Isolation": 

> snapshot isolation is a guarantee that all reads made in a transaction will see a consistent snapshot of the database

Take a scenario where I use ReplacePartitions to do an update, ie "INSERT OVERWRITE table PARTITIONS (date = '$date') SELECT table.salary + 1000 FROM table WHERE date='date'", once I read a snapshot of the row for modification, I should not be updating it if it has been modified since (ie, if we see a possibly related delete file or deleted data file in our case when we commit).

I think new data files is fine, as once we choose the set of rows to delete, they will not affect those.

What do you think?

szehon-ho commented:
@RussellSpitzer @flyrain this is the issue we talked about, if any thoughts on what is 'snapshot' isolation once we define 'serializable' isolation.

RussellSpitzer commented:
> @RussellSpitzer @flyrain this is the issue we talked about, if any thoughts on what is 'snapshot' isolation once we define 'serializable' isolation.

https://jepsen.io/consistency // Jepsen's Doc provides a good overview
https://jepsen.io/consistency/models/serializable 


Repeatable Reads is a bit different than snapshot Isolation
https://jepsen.io/consistency/models/repeatable-read
```Repeatable read is closely related to [serializability](https://jepsen.io/consistency/models/serializable), but unlike serializable, it allows [phantoms](http://pmg.csail.mit.edu/papers/adya-phd.pdf): if a transaction T1 reads a predicate, like "the set of all people with the name “Dikembe”, then another transaction T2 may create or modify a person with the name “Dikembe” before T1 commits. Individual objects are stable once read, but the predicate itself may not be.```
 
https://jepsen.io/consistency/models/snapshot-isolation
```In a snapshot isolated system, each transaction appears to operate on an independent, consistent snapshot of the database. Its changes are visible only to that transaction until commit time, when all changes become visible atomically. If transaction T1 has modified an object x, and another transaction T2 committed a write to x after T1’s snapshot began, and before T1’s commit, then T1 must abort.```

I'm pretty sure Postgres's definition there actually fits Snapshot Isolation better than Repeatable Read since [they don't allow Phantom Reads](https://www.postgresql.org/docs/13/transaction-iso.html#MVCC-ISOLEVEL-TABLE) which are allowed at a true Repeatable Reads Isolation . Snapshot Isolation in my mind says, you cannot modify records that were changed by a previous operation but you may modify records that were not modified by a previous commit while ignore the changes that previous commit produced.

Stolen from [Sql-Server Blog](https://techcommunity.microsoft.com/t5/sql-server-blog/serializable-vs-snapshot-isolation-level/ba-p/383281)
![image](https://user-images.githubusercontent.com/413025/155622462-2a862637-a089-41fd-ac73-2032de7875fd.png)

Here you imagine two commits as 
"UPDATE color=white WHERE color = BLACK" 
"UPDATE color= black where color = WHITE"

Both of these commits are allowed to occur as if they applied to the same original commit because each operation only effected an isolated set of marbles.

So in this case I believe INSERT OVERWRITE would never conflict with another INSERT, but would conflict with an update that changed any row within the partition being over-written. Another INSERT OVERWRITE would be a form of update so I believe @szehon-ho has the right of it here.

aokolnychyi commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -244,4 +245,18 @@ public boolean useTableDistributionAndOrdering() {
+        .parseOptional();
+  }
+
+  public IsolationLevel dynamicOverwriteIsolationLevel() {
```
Based on the discussion we had on isolation levels, I think this the default should be `null` (i.e. no validation). Serializable isolation should mean validate both new data files and all types of deletes. Snapshot isolation would mean validate just all types of deletes.

szehon-ho replied:
Yep, done.

aokolnychyi commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteOptions.java:
```diff
@@ -68,4 +68,11 @@ private SparkWriteOptions() {
+
+  // Isolation Level for DataFrame calls.  Currently supported by overwritePartitions
+  public static final String ISOLATION_LEVEL = "isolation-level";
+  public static final String DYNAMIC_OVERWRITE_LEVEL_DEFAULT = "snapshot";
```
Like I said above, I think the default should be no validation. I'd remove this constant.

szehon-ho replied:
Done

aokolnychyi reviewed (commented):
I agree with Russell's analysis. In my own words, I think snapshot isolation should validate concurrent deletes while serializable isolation should validate both new data files and all types of concurrent deletes. This definition matches the existing behavior of copy-on-write operations.

szehon-ho commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java:
```diff
@@ -111,6 +111,8 @@
   private final boolean partitionedFanoutEnabled;
   private final Distribution requiredDistribution;
   private final SortOrder[] requiredOrdering;
+  private final Long validateFromSnapshotId;
```
Done (though did not change the other variables, that can now be simplified).

szehon-ho commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteOptions.java:
```diff
@@ -68,4 +68,11 @@ private SparkWriteOptions() {
+  // Identifies snapshot from which to start validating conflicting changes
+  public static final String VALIDATE_FROM_SNAPSHOT_ID = "validate-from-snapshot-id";
+
+  // Isolation Level for DataFrame calls.  Currently supported by overwritePartitions
```
I must be old,  writing with two periods between sentence, changed.

szehon-ho commented on api/src/main/java/org/apache/iceberg/ReplacePartitions.java:
```diff
@@ -49,4 +55,52 @@
+   * @param snapshotId a snapshot ID, it should be set to when this operation started to read the table.
+   * @return this for method chaining
+   */
+  ReplacePartitions validateFromSnapshot(Long snapshotId);
```
Made the necessary changes

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -382,21 +429,142 @@ private void validateNoNewDeletesForDataFiles(TableMetadata base, Long startingS
+    }
+
+    Pair<List<ManifestFile>, Set<Long>> history =
+        validationHistory(base, startingSnapshotId, VALIDATE_ADDED_FILES_OPERATIONS, ManifestContent.DATA);
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -382,21 +429,142 @@ private void validateNoNewDeletesForDataFiles(TableMetadata base, Long startingS
+   */
+  private CloseableIterable<ManifestEntry<DataFile>> deletedDataFiles(TableMetadata base,
+                                                                    Long startingSnapshotId,
+                                                                    Expression detectionFilter,
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -382,21 +429,142 @@ private void validateNoNewDeletesForDataFiles(TableMetadata base, Long startingS
+   * @param partitionSet a set of partitions to find deleted data files
+   */
+  private CloseableIterable<ManifestEntry<DataFile>> deletedDataFiles(TableMetadata base,
+                                                                    Long startingSnapshotId,
```
Fixed

szehon-ho commented on core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:
```diff
@@ -272,35 +297,61 @@ private ManifestFile copyManifest(ManifestFile manifest) {
+   */
+  private CloseableIterable<ManifestEntry<DataFile>> addedDataFiles(TableMetadata base,
+                                                                    Long startingSnapshotId,
+                                                                    Expression detectionFilter,
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/ManifestReader.java:
```diff
@@ -170,7 +177,8 @@ public PartitionSpec spec() {
     if ((rowFilter != null && rowFilter != Expressions.alwaysTrue()) ||
-        (partFilter != null && partFilter != Expressions.alwaysTrue())) {
+        (partFilter != null && partFilter != Expressions.alwaysTrue()) ||
+        (partitionSet != null && !partitionSet.isEmpty())) {
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:
```diff
@@ -53,6 +63,53 @@ public ReplacePartitions validateAppendOnly() {
+  }
+
+  @Override
+  public ReplacePartitions validateNoConflictingDeletedData() {
```
Done

szehon-ho commented:
@aokolnychyi thanks for the review, addressed the comments and implemented snapshot isolation + tests based on the general agreement.  Will also change the name of the pr to reflect

aokolnychyi commented:
Thanks, @szehon-ho! This is awesome work! Thanks others for reviewing too.
